mailqueue.py

```diff
@@ -0,0 +1,213 @@
+            'message': email.message_from_string,
+            'next_attempt_at': int,
+            'recipients': lambda x: x.split(','),
+            'status': as_is,
+            'submission_id': as_is
+        }
+        return Envelope(id=row['rowid'],
+                        **{k: f(row[k]) for k, f in column_transformations.items()})
+
+    def _execute_committing(self, statement, *extra_args):
+        self._db_cursor.execute(statement, *extra_args)
+        self._db_conn.commit()
+
+    def _initialize_database(self):
+        status_column = "status TEXT CHECK({}) NOT NULL, ".format(' or '.join(
+            'status="{}"'.format(v) for k, v in Status.__dict__.items() if not k.startswith('_'))
+        )
+        self._execute_committing(
+            "CREATE TABLE IF NOT EXISTS envelopes ("
+            "recipients TEXT NOT NULL, "
+            "destination_domain TEXT NOT NULL, "
+            "message TEXT NOT NULL, "
+            "next_attempt_at INTEGER NOT NULL, "
+            + status_column +
+            "submission_id TEXT NOT NULL, "
+            "delivery_attempts INTEGER NOT NULL DEFAULT 0, "
+            "being_processed BOOLEAN NOT NULL DEFAULT 0, "
+            "client_id TEXT NOT NULL"
+            ")"
+        )
+
+    def _select_row_for_processing(self):
+        self._db_cursor.execute(
+            "SELECT rowid, * FROM envelopes "
+            "WHERE next_attempt_at <= ? AND status='{}' AND being_processed=0 "
+            "AND (rowid % ?) = ? LIMIT 1".format(Status.QUEUED),
+            (self.clock.time(), self._shards, self._shard)
+        )
+        row = self._db_cursor.fetchone()
+
+        if row:
+            self._execute_committing("UPDATE envelopes SET being_processed=1 WHERE rowid=?",
+                                     (row['rowid'],))
+        return row
+
+    def _set_envelope_status(self, envelope, status):
+        self._assert_envelope_has_id(envelope)
+        self._execute_committing(
+            'UPDATE envelopes SET status="{}" WHERE rowid=?'.format(status), (envelope.id, ))
+
+
+# TODO: rename it to SynchronizedMailQueue
+class Manager:
+    """A proxy synchronizing access to MailQueue from multiple threads.
+
+    SQLite requires that the thread creating the database object is
+    the only one using it. This manager synchronizes access to methods
+    of MailQueue via blocking queues.
+    """
+
+    def __init__(self, **kwargs):
+        self._mail_queue_args = kwargs
+        self._manager_thread = threading.Thread(target=self._main_loop)
+        self._requests = queue.Queue(1)
+        self._responses = queue.Queue(1)
+
+        # XXX: factor out
+        methods = [name for name, _ in inspect.getmembers(MailQueue, inspect.isfunction)
+                   if not name.startswith('_')]
+
+        for method in methods:
+            assert method not in self.__class__.__dict__, \
+                "name clash: method '{}' already exists in {}".format(method,
```

Comment:
Not a question but a general comment. This seems like a nice trick that saves some typing... but in practice the code written using this trick is very hard to maintain. You can never obviously tell what methods a class has. IDEs do not understand that and therefore code completion and code insight do not work. Grep returns incomplete results... So even if we do not extend our offer please do not ever do that for your colleagues sake :)

Reply:
I agree, this is not as straightforward as I would like. However, this is not to save typing.

The tradeoff here is this:

Option A: keep it simple, and manually keep parity between MailQueue and Manager's methods. Once you forget to mirror a method, this will only be discovered in runtime (Manager runs in its own thread, which makes it more difficult to test).

Option B: go with the more complex solution. Accept the need for higher mental effort to comprehend the code, forfeit IDE support.  Manager always mirrors MailQueue.

I started with option A, but as the number of methods in MailQueue grew, I found myself catching bugs of missing methods in end-to-end tests. At this stage (and in real life it would be CI/CD pipeline or staging, or production) bugs are much more expensive to fix, so I opted for more complex and less readable solution to gain in correctness and feedback time.

I think my colleagues would appreciate not being woken up during the night to find out I forgot to add a method to Manager. ;-)

So I agree with the general comment, but in this case I judged it was worth it (for me). If the team decides otherwise, the other option can always be implemented.

Reply:
In all honesty I've been there and I'd rather wake up, fix the missing method and go back to sleep. :-)

Reply:
All right, I've reimplemented Manager with explicitly defined methods:

https://github.com/wadcom/mini-mailgun/pull/3/commits/6f08849eacf268f3aa4331a65632619521b42456